File: scripts/CheckToolVersions.java
/// Check that the dev image's pinned tools are still the latest upstream release.
///
/// Reads each `ARG <TOOL>_VERSION=...` pin from the Dockerfile and compares it with
/// the newest matching GitHub release. A single-file Java program — run it directly
/// with `java scripts/CheckToolVersions.java` (no build step); it needs only a JDK,
/// which the dev image and CI already provide. The apt-installed tools (qpdf,
/// poppler-utils, webp) track the Ubuntu base and are not pinned here, so they are
/// intentionally not checked.
///
/// Exit status is 1 if any tool is behind, 2 if a lookup fails, 0 if all current.
/// Set GITHUB_TOKEN to raise the GitHub API rate limit (optional). Run from the repo
/// root (where the Dockerfile lives).
import java.io.IOException;
import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.nio.file.Files;
import java.nio.file.Path;
import java.time.Duration;
import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

public final class CheckToolVersions {

    /** A pinned tool: its Dockerfile ARG, GitHub repo, and a tag pattern whose group 1 is the version. */
    record Tool(String arg, String repo, Pattern tag) {}

    // Most tags are vX.Y.Z (an optional leading v); biome tags are @biomejs/biome@X.Y.Z.
    private static final Pattern V_TAG = Pattern.compile("^v?(\\d.+)$");

    private static final List<Tool> TOOLS =
            List.of(
                    new Tool("JUST_VERSION", "casey/just", V_TAG),
                    new Tool("LEFTHOOK_VERSION", "evilmartians/lefthook", V_TAG),
                    new Tool("TYPOS_VERSION", "crate-ci/typos", V_TAG),
                    new Tool("TAPLO_VERSION", "tamasfe/taplo", V_TAG),
                    new Tool("BIOME_VERSION", "biomejs/biome", Pattern.compile("^@biomejs/biome@(\\d.+)$")),
                    new Tool("YAMLFMT_VERSION", "google/yamlfmt", V_TAG),
                    new Tool("ACTIONLINT_VERSION", "rhysd/actionlint", V_TAG),
                    new Tool("JBIG2ENC_VERSION", "agl/jbig2enc", V_TAG));

    private static final Pattern DIGITS = Pattern.compile("\\d+");

    private static final HttpClient HTTP =
            HttpClient.newBuilder().connectTimeout(Duration.ofSeconds(30)).build();

    public static void main(String[] args) throws Exception {
        Map<String, String> pins = pinnedVersions(Path.of("Dockerfile"));
        int outdated = 0;
        int errors = 0;

        System.out.printf("%-20s %-12s %-12s %s%n", "tool", "pinned", "latest", "status");
        System.out.println("-".repeat(58));

        for (Tool tool : TOOLS) {
            String pinned = pins.getOrDefault(tool.arg(), "?");
            String latest;
            try {
                latest = latestVersion(tool);
            } catch (IOException | InterruptedException e) {
                System.out.printf(
                        "%-20s %-12s %-12s %s%n",
                        tool.arg(), pinned, "?", "ERROR (" + tool.repo() + "): " + e.getMessage());
                errors++;
                continue;
            }
            if (latest == null) {
                System.out.printf(
                        "%-20s %-12s %-12s %s%n",
                        tool.arg(), pinned, "?", "no matching release (" + tool.repo() + ")");
                errors++;
            } else if (compareKey(versionKey(latest), versionKey(pinned)) > 0) {
                System.out.printf(
                        "%-20s %-12s %-12s %s%n",
                        tool.arg(), pinned, latest, "OUTDATED (" + tool.repo() + ")");
                outdated++;
            } else {
                System.out.printf("%-20s %-12s %-12s %s%n", tool.arg(), pinned, latest, "ok");
            }
        }

        System.out.println();
        if (outdated > 0) {
            System.out.println(outdated + " tool(s) behind latest — bump the ARG in the Dockerfile.");
        }
        if (errors > 0) {
            System.out.println(errors + " tool(s) could not be checked.");
        }
        if (outdated == 0 && errors == 0) {
            System.out.println("all pinned tools are at the latest release.");
        }
        System.exit(outdated > 0 ? 1 : (errors > 0 ? 2 : 0));
    }

    /** The {@code ARG <name>=<value>} pins for the known tools, read from the Dockerfile. */
    private static Map<String, String> pinnedVersions(Path dockerfile) throws IOException {
        String text = Files.readString(dockerfile);
        Map<String, String> pins = new LinkedHashMap<>();
        for (Tool tool : TOOLS) {
            Matcher m =
                    Pattern.compile("^ARG " + tool.arg() + "=(.+)$", Pattern.MULTILINE).matcher(text);
            if (m.find()) {
                pins.put(tool.arg(), m.group(1).strip());
            }
        }
        return pins;
    }

    /** The newest non-draft, non-prerelease release version matching the tool's tag pattern, or null. */
    private static String latestVersion(Tool tool) throws IOException, InterruptedException {
        Object parsed =
                Json.parse(
                        apiGet(
                                "https://api.github.com/repos/"
                                        + tool.repo()
                                        + "/releases?per_page=100"));
        if (!(parsed instanceof List<?> releases)) {
            return null;
        }
        String best = null;
        for (Object element : releases) {
            if (!(element instanceof Map<?, ?> release)) {
                continue;
            }
            if (Boolean.TRUE.equals(release.get("draft"))
                    || Boolean.TRUE.equals(release.get("prerelease"))) {
                continue;
            }
            if (!(release.get("tag_name") instanceof String tagName)) {
                continue;
            }
            Matcher m = tool.tag().matcher(tagName);
            if (m.matches()
                    && (best == null || compareKey(versionKey(m.group(1)), versionKey(best)) > 0)) {
                best = m.group(1);
            }
        }
        return best;
    }

    private static String apiGet(String url) throws IOException, InterruptedException {
        HttpRequest.Builder request =
                HttpRequest.newBuilder(URI.create(url))
                        .header("Accept", "application/vnd.github+json")
                        .header("User-Agent", "despeckle-tool-version-check")
                        .timeout(Duration.ofSeconds(30));
        String token = System.getenv("GITHUB_TOKEN");
        if (token != null && !token.isBlank()) {
            request.header("Authorization", "Bearer " + token);
        }
        HttpResponse<String> response =
                HTTP.send(request.build(), HttpResponse.BodyHandlers.ofString());
        if (response.statusCode() != 200) {
            throw new IOException("HTTP " + response.statusCode());
        }
        return response.body();
    }

    /** The integer runs in a version, compared lexicographically (1.10 > 1.9). */
    private static int[] versionKey(String version) {
        Matcher m = DIGITS.matcher(version);
        List<Integer> parts = new ArrayList<>();
        while (m.find()) {
            parts.add(Integer.parseInt(m.group()));
        }
        return parts.stream().mapToInt(Integer::intValue).toArray();
    }

    private static int compareKey(int[] a, int[] b) {
        for (int i = 0; i < Math.min(a.length, b.length); i++) {
            if (a[i] != b[i]) {
                return Integer.compare(a[i], b[i]);
            }
        }
        return Integer.compare(a.length, b.length);
    }

    private CheckToolVersions() {}

    /**
     * A tiny recursive-descent JSON reader — just enough to walk GitHub's releases array (objects,
     * arrays, strings, numbers, booleans, null). The JDK ships no JSON parser, and a single-file
     * program has only the JDK on its classpath, so this stays self-contained.
     */
    static final class Json {
        private final String src;
        private int pos;

        private Json(String src) {
            this.src = src;
        }

        static Object parse(String src) {
            Json json = new Json(src);
            json.skipWhitespace();
            return json.readValue();
        }

        private Object readValue() {
            skipWhitespace();
            char c = src.charAt(pos);
            return switch (c) {
                case '{' -> readObject();
                case '[' -> readArray();
                case '"' -> readString();
                case 't', 'f' -> readBoolean();
                case 'n' -> readNull();
                default -> readNumber();
            };
        }

        private Map<String, Object> readObject() {
            Map<String, Object> map = new LinkedHashMap<>();
            expect('{');
            skipWhitespace();
            if (src.charAt(pos) == '}') {
                pos++;
                return map;
            }
            while (true) {
                skipWhitespace();
                String key = readString();
                skipWhitespace();
                expect(':');
                map.put(key, readValue());
                skipWhitespace();
                char c = src.charAt(pos++);
                if (c == '}') {
                    return map;
                }
                if (c != ',') {
                    throw error("expected ',' or '}'");
                }
            }
        }

        private List<Object> readArray() {
            List<Object> list = new ArrayList<>();
            expect('[');
            skipWhitespace();
            if (src.charAt(pos) == ']') {
                pos++;
                return list;
            }
            while (true) {
                list.add(readValue());
                skipWhitespace();
                char c = src.charAt(pos++);
                if (c == ']') {
                    return list;
                }
                if (c != ',') {
                    throw error("expected ',' or ']'");
                }
            }
        }

        private String readString() {
            expect('"');
            StringBuilder sb = new StringBuilder();
            while (true) {
                char c = src.charAt(pos++);
                if (c == '"') {
                    return sb.toString();
                }
                if (c != '\\') {
                    sb.append(c);
                    continue;
                }
                char escape = src.charAt(pos++);
                switch (escape) {
                    case '"' -> sb.append('"');
                    case '\\' -> sb.append('\\');
                    case '/' -> sb.append('/');
                    case 'b' -> sb.append('\b');
                    case 'f' -> sb.append('\f');
                    case 'n' -> sb.append('\n');
                    case 'r' -> sb.append('\r');
                    case 't' -> sb.append('\t');
                    case 'u' -> {
                        sb.append((char) Integer.parseInt(src.substring(pos, pos + 4), 16));
                        pos += 4;
                    }
                    default -> throw error("bad escape \\" + escape);
                }
            }
        }

        private Boolean readBoolean() {
            if (src.startsWith("true", pos)) {
                pos += 4;
                return Boolean.TRUE;
            }
            if (src.startsWith("false", pos)) {
                pos += 5;
                return Boolean.FALSE;
            }
            throw error("bad literal");
        }

        private Object readNull() {
            if (src.startsWith("null", pos)) {
                pos += 4;
                return null;
            }
            throw error("bad literal");
        }

        private Double readNumber() {
            int start = pos;
            while (pos < src.length() && "+-0123456789.eE".indexOf(src.charAt(pos)) >= 0) {
                pos++;
            }
            return Double.valueOf(src.substring(start, pos));
        }

        private void skipWhitespace() {
            while (pos < src.length() && Character.isWhitespace(src.charAt(pos))) {
                pos++;
            }
        }

        private void expect(char c) {
            if (src.charAt(pos++) != c) {
                throw error("expected '" + c + "'");
            }
        }

        private IllegalArgumentException error(String message) {
            return new IllegalArgumentException("JSON: " + message + " at offset " + pos);
        }
    }
}

File: scripts/check-tool-versions.py
#!/usr/bin/env python3
"""Check that the dev image's pinned tools are still the latest upstream release.

Reads each `ARG <TOOL>_VERSION=...` pin from the Dockerfile and compares it with
the newest matching GitHub release. The apt-installed tools (qpdf, poppler-utils,
webp) are not pinned here — they track the Ubuntu base and refresh on every image
build — so they are intentionally not checked.

Exit status is non-zero if any tool is behind (so `just tools-latest` and the
scheduled CI job both go red when something needs bumping) or if a lookup fails.

Set GITHUB_TOKEN to raise the API rate limit (optional).
"""
import json
import os
import re
import sys
import urllib.error
import urllib.request

# ARG name -> (GitHub repo, regex whose first group is the version in a release tag).
TOOLS = {
    "JUST_VERSION": ("casey/just", r"^v?(\d.+)$"),
    "LEFTHOOK_VERSION": ("evilmartians/lefthook", r"^v?(\d.+)$"),
    "TYPOS_VERSION": ("crate-ci/typos", r"^v?(\d.+)$"),
    "TAPLO_VERSION": ("tamasfe/taplo", r"^v?(\d.+)$"),
    "BIOME_VERSION": ("biomejs/biome", r"^@biomejs/biome@(\d.+)$"),
    "YAMLFMT_VERSION": ("google/yamlfmt", r"^v?(\d.+)$"),
    "ACTIONLINT_VERSION": ("rhysd/actionlint", r"^v?(\d.+)$"),
    "JBIG2ENC_VERSION": ("agl/jbig2enc", r"^v?(\d.+)$"),
}

DOCKERFILE = os.path.join(os.path.dirname(__file__), os.pardir, "Dockerfile")


def pinned_versions():
    with open(DOCKERFILE, encoding="utf-8") as f:
        text = f.read()
    pins = {}
    for arg in TOOLS:
        match = re.search(rf"^ARG {arg}=(.+)$", text, re.MULTILINE)
        if match:
            pins[arg] = match.group(1).strip()
    return pins


def api_get(url):
    request = urllib.request.Request(
        url,
        headers={
            "Accept": "application/vnd.github+json",
            "User-Agent": "despeckle-tool-version-check",
        },
    )
    token = os.environ.get("GITHUB_TOKEN")
    if token:
        request.add_header("Authorization", f"Bearer {token}")
    with urllib.request.urlopen(request, timeout=30) as response:
        return json.load(response)


def version_key(version):
    return [int(part) for part in re.findall(r"\d+", version)]


def latest_version(repo, pattern):
    """Newest non-draft, non-prerelease release tag matching pattern, or None."""
    releases = api_get(f"https://api.github.com/repos/{repo}/releases?per_page=100")
    best = None
    for release in releases:
        if release.get("draft") or release.get("prerelease"):
            continue
        match = re.match(pattern, release.get("tag_name", ""))
        if match and (best is None or version_key(match.group(1)) > version_key(best)):
            best = match.group(1)
    return best


def main():
    pins = pinned_versions()
    outdated = errors = 0
    print(f"{'tool':<20} {'pinned':<10} {'latest':<10} status")
    print("-" * 54)
    for arg, (repo, pattern) in TOOLS.items():
        pinned = pins.get(arg, "?")
        try:
            latest = latest_version(repo, pattern)
        except urllib.error.HTTPError as exc:
            print(f"{arg:<20} {pinned:<10} {'?':<10} ERROR {exc.code} ({repo})")
            errors += 1
            continue
        if latest is None:
            print(f"{arg:<20} {pinned:<10} {'?':<10} no matching release ({repo})")
            errors += 1
        elif version_key(latest) > version_key(pinned):
            print(f"{arg:<20} {pinned:<10} {latest:<10} OUTDATED ({repo})")
            outdated += 1
        else:
            print(f"{arg:<20} {pinned:<10} {latest:<10} ok")
    print()
    if outdated:
        print(f"{outdated} tool(s) behind latest — bump the ARG in the Dockerfile.")
    if errors:
        print(f"{errors} tool(s) could not be checked.")
    if not outdated and not errors:
        print("all pinned tools are at the latest release.")
    return 1 if outdated else (2 if errors else 0)


if __name__ == "__main__":
    sys.exit(main())
